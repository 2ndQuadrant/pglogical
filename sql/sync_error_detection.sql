/* Test for sync error handling in RM#2829 */

SELECT * FROM pglogical_regress_variables()
\gset

\c :provider_dsn

\c :subscriber_dsn

-- Create a downstream-only table then try to resync it from the provider. This
-- will fail since the provider table does not exist.

CREATE TABLE public.test_sync_error(id serial primary key, data text);

SELECT * FROM pglogical.alter_subscription_resynchronize_table('test_subscription', 'test_sync_error');

DO $$
-- Wait until the sync has started
BEGIN
	FOR i IN 1..100 LOOP
		IF (SELECT count(1) FROM pglogical.local_sync_status WHERE sync_status != 'i' AND sync_relname = 'test_sync_error') = 1 THEN
			RETURN;
		END IF;
		PERFORM pg_sleep(0.1);
	END LOOP;
END;
$$;

-- sync_status will always be an in-progress state here, because the worker exits
-- without changing it, and a new worker will set it again.
SELECT sync_kind, sync_subid, sync_nspname, sync_relname, sync_status IN ('r', 'y') AS sync_ok, sync_status NOT IN ('y', 'r') AS sync_in_progress
FROM pglogical.local_sync_status
WHERE sync_relname = 'test_sync_error'
ORDER BY 2,3,4;

-- If we create the table on the upstream and ask to resync again it should work

\c :provider_dsn

CREATE TABLE public.test_sync_error(id serial primary key, data text);

INSERT INTO public.test_sync_error(data) VALUES ('awk awk awk');

SELECT pglogical.wait_slot_confirm_lsn(NULL, NULL);

\c :subscriber_dsn

-- The new row must not have applied, since the table is in failed-sync state and not receiving changes
SELECT * FROM pglogical.test_sync_error;

-- But we can resync it now. Add some local content first to make sure it gets clobbered.
INSERT INTO public.test_sync_error(data) VALUES ('dead parrot');

SELECT * FROM pglogical.alter_subscription_resynchronize_table('test_subscription', 'test_sync_error');

DO $$
-- give it 10 seconds to syncrhonize the tabes
BEGIN
	FOR i IN 1..100 LOOP
		IF (SELECT count(1) FROM pglogical.local_sync_status WHERE sync_status IN ('y', 'r') AND sync_relname = 'test_sync_error') = 1 THEN
			RETURN;
		END IF;
		PERFORM pg_sleep(0.1);
	END LOOP;
END;
$$;

SELECT sync_kind, sync_subid, sync_nspname, sync_relname, sync_status IN ('r', 'y') AS sync_ok
FROM pglogical.local_sync_status
WHERE sync_relname = 'test_sync_error'
ORDER BY 2,3,4;

\c :provider_dsn

\set VERBOSITY terse
SELECT pglogical.replicate_ddl_command($$
	DROP TABLE public.test_sync_error;
$$);
