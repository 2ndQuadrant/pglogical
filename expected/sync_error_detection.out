/* Test for sync error handling in RM#2829 */
SELECT * FROM pglogical_regress_variables()
\gset
\c :provider_dsn
\c :subscriber_dsn
-- Create a downstream-only table then try to resync it from the provider. This
-- will fail since the provider table does not exist.
CREATE TABLE public.test_sync_error(id serial primary key, data text);
SELECT * FROM pglogical.alter_subscription_resynchronize_table('test_subscription', 'test_sync_error');
 alter_subscription_resynchronize_table 
----------------------------------------
 t
(1 row)

DO $$
-- Wait until the sync has started
BEGIN
	FOR i IN 1..100 LOOP
		IF (SELECT count(1) FROM pglogical.local_sync_status WHERE sync_status != 'i' AND sync_relname = 'test_sync_error') = 1 THEN
			RETURN;
		END IF;
		PERFORM pg_sleep(0.1);
	END LOOP;
END;
$$;
-- sync_status will always be an in-progress state here, because the worker exits
-- without changing it, and a new worker will set it again.
SELECT sync_kind, sync_subid, sync_nspname, sync_relname, sync_status IN ('r', 'y') AS sync_ok, sync_status NOT IN ('y', 'r') AS sync_in_progress
FROM pglogical.local_sync_status
WHERE sync_relname = 'test_sync_error'
ORDER BY 2,3,4;
 sync_kind | sync_subid | sync_nspname |  sync_relname   | sync_ok | sync_in_progress 
-----------+------------+--------------+-----------------+---------+------------------
 d         | 3848008564 | public       | test_sync_error | f       | t
(1 row)

-- If we create the table on the upstream and ask to resync again it should work
\c :provider_dsn
CREATE TABLE public.test_sync_error(id serial primary key, data text);
INSERT INTO public.test_sync_error(data) VALUES ('awk awk awk');
SELECT pglogical.wait_slot_confirm_lsn(NULL, NULL);
 wait_slot_confirm_lsn 
-----------------------
 
(1 row)

\c :subscriber_dsn
-- The new row must not have applied, since the table is in failed-sync state and not receiving changes
SELECT * FROM pglogical.test_sync_error;
ERROR:  relation "pglogical.test_sync_error" does not exist
LINE 1: SELECT * FROM pglogical.test_sync_error;
                      ^
-- But we can resync it now. Add some local content first to make sure it gets clobbered.
INSERT INTO public.test_sync_error(data) VALUES ('dead parrot');
ERROR:  duplicate key value violates unique constraint "test_sync_error_pkey"
DETAIL:  Key (id)=(1) already exists.
SELECT * FROM pglogical.alter_subscription_resynchronize_table('test_subscription', 'test_sync_error');
ERROR:  table public.test_sync_error is already being synchronized
DO $$
-- give it 10 seconds to syncrhonize the tabes
BEGIN
	FOR i IN 1..100 LOOP
		IF (SELECT count(1) FROM pglogical.local_sync_status WHERE sync_status IN ('y', 'r') AND sync_relname = 'test_sync_error') = 1 THEN
			RETURN;
		END IF;
		PERFORM pg_sleep(0.1);
	END LOOP;
END;
$$;
SELECT sync_kind, sync_subid, sync_nspname, sync_relname, sync_status IN ('r', 'y') AS sync_ok
FROM pglogical.local_sync_status
WHERE sync_relname = 'test_sync_error'
ORDER BY 2,3,4;
 sync_kind | sync_subid | sync_nspname |  sync_relname   | sync_ok 
-----------+------------+--------------+-----------------+---------
 d         | 3848008564 | public       | test_sync_error | t
(1 row)

\c :provider_dsn
\set VERBOSITY terse
SELECT pglogical.replicate_ddl_command($$
	DROP TABLE public.test_sync_error;
$$);
 replicate_ddl_command 
-----------------------
 t
(1 row)

